### ISSUE TITLE
Closing dialog with Escape

### ISSUE THREAD

**Issue by exectails** [aura-project/melia#13]
When pressing escape, an open dialog should be closed. Right now we're resuming and continuing the dialog, which is not official behavior and can get the player stuck in the dialog until relog.

The unknown int in `CZ_DIALOG_ACK` seems to indicate what happened, with 0 or -1 being sent when Escape was pressed and 1 otherwise.

**Comment by exectails** [comment]
I'm not sure how to do this right now, as simply closing the dialog on the client when pressing escape doesn't stop the yielding Lua thread. If you talk to the NPC again it will continue where it left off. We have to cancel the yield. We could create a new thread for the client to use, but then we have yield hanging around. I don't know if Lua would collect that.

**Comment by exectails** [comment]
A potential solution could be to not use lua_yield anymore, but C#'s async features, to suspend the operation from within the ScriptManager, similar to how we did stateless NPCs in Aura.

https://github.com/aura-project/aura/blob/master/src/ChannelServer/Scripting/Scripts/NpcScript.cs#L2048

That way we don't leave the suspending function and might be able to stop the execution, and be it by returning an error.

**Comment by exectails** [comment]
I did some research on how the Lua threads work and figured out how to solve all my Lua problems. When a new thread is created with `lua_newthread` it's put on the global state's stack. It stays there until it's removed, and only after that it's collected. This means we simply have to remove the thread from the stack to clean everything up.

For more flexibility I also changed the way threads are created, so it's one thread per *dialog sessions*, instead of play session.
